### PR TITLE
[SPARK-23754][PYTHON][FOLLOWUP] Move UDF stop iteration wrapping from driver to executor

### PR DESCRIPTION
**Pull request by e-dorigatti**
## What changes were proposed in this pull request?
SPARK-23754 was fixed in #21383 by changing the UDF code to wrap the user function, but this required a hack to save its argspec. This PR reverts this change and fixes the `StopIteration` bug in the worker

## How does this work?

The root of the problem is that when an user-supplied function raises a `StopIteration`, pyspark might stop processing data, if this function is used in a for-loop. The solution is to catch `StopIteration`s exceptions and re-raise them as `RuntimeError`s, so that the execution fails and the error is reported to the user. This is done using the `fail_on_stopiteration` wrapper, in different ways depending on where the function is used:
 - In RDDs, the user function is wrapped in the driver, because this function is also called in the driver itself.
 - In SQL UDFs, the function is wrapped in the worker, since all processing happens there. Moreover, the worker needs the signature of the user function, which is lost when wrapping it, but passing this signature to the worker requires a not so nice hack.

## How was this patch tested?

Same tests, plus tests for pandas UDFs

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
add to whitelist

**Comment by e-dorigatti** [comment]
Seems like a problem in Jenkins

```
ImportError: PyArrow >= 0.8.0 must be installed; however, it was not found.
```

File: python/pyspark/sql/tests.py
Comment (icexelloss) [inline python/pyspark/sql/tests.py]: Can you please add a tests for grouped_map with 2 args?

File: python/pyspark/worker.py
Comment (icexelloss) [inline python/pyspark/worker.py]: Should we move the wrapping to `read_single_udf` instead of putting in individual `wrap_*` functions?

File: python/pyspark/sql/tests.py
Comment (icexelloss) [inline python/pyspark/sql/tests.py]: I think you can use `self.assertRaisesRegex` instead

File: python/pyspark/sql/tests.py
Comment (icexelloss) [inline python/pyspark/sql/tests.py]: apply.show()?

File: python/pyspark/sql/tests.py
Comment (icexelloss) [inline python/pyspark/sql/tests.py]: Missing show()

File: python/pyspark/worker.py
Comment (e-dorigatti) [reply python/pyspark/worker.py]: `wrap_grouped_map_pandas_udf` needs to call `argspec`, so we would need to check `evalType` twice in `read_single_udf`. But it's possible, yea

**Comment by icexelloss** [comment]
Thanks @e-dorigatti . This looks much cleaner.

Can you rename the title to:
[SPARK-23754][PYTHON][FOLLOWUP] Move UDF stop iteration wrapping from driver to executor

File: python/pyspark/worker.py
Comment (e-dorigatti) [reply python/pyspark/worker.py]: Or just move the call to `get_argspec` in `read_single_udf`

Comment (icexelloss) [reply python/pyspark/worker.py]: Can we wrap `fail_on_stopiteration` after `wrap*_udf`?

Comment (icexelloss) [reply python/pyspark/worker.py]: >move the call to get_argspec in read_single_udf

Yeah I think that works

**Comment by viirya** [comment]
hmm, should we revert #21383 and then do the new change? Is it better for review?

**Comment by HyukjinKwon** [comment]
Eh, it shouldn't be difficult to review, I guess. I think there's a mistake.

File: python/pyspark/flycheck_worker.py
Comment (HyukjinKwon) [inline python/pyspark/flycheck_worker.py]: Seems mistakenly added ..?

**Comment by HyukjinKwon** [comment]
@e-dorigatti BTW, let's minimise the change to make the backport easier.

**Comment by HyukjinKwon** [comment]
Oh you mean revert it here in this PR. yea sure that should be easier to review.

**Comment by vanzin** [comment]
SPARK-23754 was a blocker for 2.3.1, does this mean it should be re-opened and still block the release?

**Comment by vanzin** [comment]
Ping? (I know it's the middle of the night for a few of you, but I was hoping there was somebody who could answer the above question...)

**Comment by e-dorigatti** [comment]
I'm the least qualified to answer here, but technically the bug is fixed
and this is just to clean the code a bit, so it shouldn't be a blocker (but
don't quote me on that)

**Comment by vanzin** [comment]
Sounds good. Too late now for me to work on the next rc, but I'll assume that it's ok to do that unless I hear otherwise before tomorrow morning.

File: python/pyspark/flycheck_worker.py
Comment (viirya) [reply python/pyspark/flycheck_worker.py]: I don't see where it is used.

**Comment by HyukjinKwon** [comment]
Yup, I merged it to unblock the release. This PR just targets to clean up.

**Comment by e-dorigatti** [comment]
@viirya we only want to reverd `udf.py` and the hack in `_get_argspec`. Did I miss anything there?

**Comment by HyukjinKwon** [comment]
@e-dorigatti, can you maybe rebase and then start this with revert commit? That should make guys less confused. I would do squash commits first, revert it and then bring the squashed commit back. For example, `git reset HEAD^`, `git stash`, `git revert 0ebb0c0d4dd3e192464dc5e0e6f01efa55b945ed`, `git stash pop` with manual conflict resolution and `git push origin fix_udf_hack --force`.

**Comment by e-dorigatti** [comment]
messed up, of course :) fixing

**Comment by e-dorigatti** [comment]
`PyArrow` again

**Comment by e-dorigatti** [comment]
@viirya for RDDs, we cannot check for stopiterations in the executor because this bug is introduced in the driver by rewriting the methods. consider rdd.map:
```
     def map(self, f, preservesPartitioning=False):
         def func(_, iterator):
-            return map(f, iterator)
+            return map(fail_on_stopiteration(f), iterator)
         return self.mapPartitionsWithIndex(func, preservesPartitioning)
```

File: python/pyspark/sql/tests.py
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: Ah, in the Jenkins, we run this tests via Python 2, Python 3 and PyPy. Python 2 and PyPy don't have PyArrow installed, so it throws an exception. It should be placed in `*PandasUDFTests` for each. For example, this is a Scalar Pandas UDF which should be placed in `ScalarPandasUDFTests`.

File: python/pyspark/sql/tests.py
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: Likewise, let's just do

```python
exc_message = "Caught StopIteration thrown from user's code; failing the task"

with self.assertRaisesRegexp(Py4JJavaError, exc_message):
    ... pandas_udf ...
```

I think it doesn't particularly reduce the codes.

File: python/pyspark/worker.py
Comment (HyukjinKwon) [inline python/pyspark/worker.py]: I think `row_func` name was fine. Let's just leave it as was.

File: python/pyspark/sql/tests.py
Comment (HyukjinKwon) [reply python/pyspark/sql/tests.py]: Oh, actually let's put them in `PandasUDFTests`.

File: python/pyspark/worker.py
Comment (e-dorigatti) [reply python/pyspark/worker.py]: I wanted to avoid the overhead of calling get_argspec even when it's not needed

Comment (HyukjinKwon) [reply python/pyspark/worker.py]: Ah, sure.

File: python/pyspark/util.py
Comment (e-dorigatti) [reply python/pyspark/util.py]: no, this is the purpose of this PR :) that's how we fixed a bug in [a previous PR](https://github.com/apache/spark/pull/21383), but we felt it was a hack so now we are doing it properly

Comment (icexelloss) [reply python/pyspark/util.py]: I meant. The comment itself can be moved back to the "else" block? (This is minor though)

**Comment by viirya** [comment]
@e-dorigatti I see. Thanks.

File: python/pyspark/util.py
Comment (viirya) [reply python/pyspark/util.py]: yea, I think the comment is for the else block.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by e-dorigatti** [comment]
does jenkins show _all_ the failed tests? or just the first one?

**Comment by HyukjinKwon** [comment]
Jenkins has 0.8.0 pyarrow and 0.19 pandas.  it's unstable feature. we should make it stable and test multiple matrixs too

**Comment by HyukjinKwon** [comment]
of course separate jira

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@e-dorigatti, just fyi you can type "retest this please" in your PR to retrigger the test since I added you in the whitelist.

**Comment by e-dorigatti** [comment]
@HyukjinKwon thank you, I though it was reserved to members

**Comment by e-dorigatti** [comment]
retest this please

File: python/pyspark/sql/tests.py
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: tiny nit: I would do:

```
self.assertRaisesRegexp(
    Py4JJavaError, exc_message, df.withColumn('v', udf(foo)('id')).collect)
```

or

```
self.assertRaisesRegexp(
    Py4JJavaError,
    exc_message,
    df.withColumn('v', udf(foo)('id')).collect)
```

**Comment by HyukjinKwon** [comment]
Hm, I thought we could do almost all of it in worker side but I just took a closer look and just understood https://github.com/apache/spark/pull/21467#issuecomment-393907314.

**Comment by viirya** [comment]
@HyukjinKwon Yeah, that is also what I thought at the first glance. :)

File: python/pyspark/tests.py
Comment (viirya) [inline python/pyspark/tests.py]: What does this mean? The exception is non-deterministic?

**Comment by HyukjinKwon** [comment]
So .. @e-dorigatti, mind if I ask to elaborate and describe the current approach within `fail_on_stopiteration`? Seems we will handle UDFs in worker side and RDD APIs (which takes a function processing each record) will be when the function is wrapped.

Let's try to explain this in details everywhere in the codes since it's going to take a while for other guys (at least me) to understand. I think I still slightly more prefer the current way to get rid of weird hacks. We already have some divergence in SQL UDF vs RDD UDF.

File: python/pyspark/tests.py
Comment (HyukjinKwon) [reply python/pyspark/tests.py]: Yea, I asked this before. He explained that the exception can be thrown in driver side or executor side too non-deterministically. We should clarify this comment. It's quite core fix. Let's clarify everything.

File: python/pyspark/worker.py
Comment (e-dorigatti) [reply python/pyspark/worker.py]: Clearer?

**Comment by e-dorigatti** [comment]
I updated the description, so that it will be included in the commit message, and slightly clarified some comments

**Review by icexelloss — APPROVED** [review]
LGTM!

**Comment by HyukjinKwon** [comment]
Seems fine. will take a look before merginf this in. Thanks for working on this.

**Comment by HyukjinKwon** [comment]
ok ti test

**Comment by HyukjinKwon** [comment]
ok to test

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Merged to master.

@e-dorigatti, it has some conflicts in branch-2.3 too. Mind if I ask to open a backporting PR again to reduce the difference between master and branch-2.3?

**Pull request closed without merging**
